src/Plugin/search_api/processor/Property/EntityReferenceWithUriProperty.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+namespace Drupal\islandora\Plugin\search_api\processor\Property;
+
+use Drupal\Core\Entity\EntityTypeManager;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\islandora\IslandoraUtils;
+use Drupal\search_api\Item\FieldInterface;
+use Drupal\search_api\Processor\ConfigurablePropertyBase;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Defines a "Entity Reference with URI" search property.
+ *
+ * @see \Drupal\islandora\Plugin\search_api\processor\EntityReferenceWithUri
+ */
+class EntityReferenceWithUriProperty extends ConfigurablePropertyBase {
+
+  use StringTranslationTrait;
+  /**
+   * Entity Type Manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManager
+   */
+  protected EntityTypeManager $entityTypeManager;
+  /**
+   * Islandora utils.
+   *
+   * @var \Drupal\islandora\IslandoraUtils
+   */
+  protected IslandoraUtils $utils;
+
+  /**
+   * Constructor.
+   *
+   * @param array $configuration
+   *   The plugin configuration, i.e. an array with configuration values keyed
+   *   by configuration option name. The special key 'context' may be used to
+   *   initialize the defined contexts by setting it to an array of context
+   *   values keyed by context names.
+   * @param \Drupal\Core\Entity\EntityTypeManager $entityTypeManager
+   *   The Entity Type Manager.
+   * @param \Drupal\islandora\IslandoraUtils $utils
+   *   Islandora utils.
+   */
+  public function __construct(
+    array $configuration,
+    EntityTypeManager $entityTypeManager,
+    IslandoraUtils $utils
+  ) {
+    parent::__construct($configuration);
+    $this->entityTypeManager = $entityTypeManager;
+    $this->utils = $utils;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration) {
```

Comment:
When enabling the processor I get:

```
Warning: Declaration of Drupal\islandora\Plugin\search_api\processor\Property\EntityReferenceWithUriProperty::create(Symfony\Component\DependencyInjection\ContainerInterface $container, array $configuration) should be compatible with Drupal\Core\TypedData\DataDefinition::create($type) in include() (line 18 of modules/contrib/islandora/src/Plugin/search_api/processor/Property/EntityReferenceWithUriProperty.php).

include('/opt/www/drupal/modules/contrib/islandora/src/Plugin/search_api/processor/Property/EntityReferenceWithUriProperty.php') (Line: 578)
Composer\Autoload\ClassLoader::Composer\Autoload\{closure}('/opt/www/drupal/modules/contrib/islandora/src/Plugin/search_api/processor/Property/EntityReferenceWithUriProperty.php') (Line: 432)
```

`search_api`'s properties are intended to be a [data definition](https://git.drupalcode.org/project/drupal/-/blob/9.4.10/core/lib/Drupal/Core/TypedData/DataDefinition.php#L28) and appear to not be architected to directly support DI. Normally when wanting to inject you would need to implement the `ContainerFactoryPluginInterface` which is not possible here.

Is a few options which could work around this such as implementing your own static factory method to instantiate the property or re-working the plugin itself to pass in the requested values directly as opposed to the property itself. Alternative could be just making static functions in the property class as opposed to injecting (@adam-vessey clarify if I'm off here?)

Reply:
I think we discussed how `ContainerFactoryPluginInterface` doesn't work here because the Property is not a Plugin, when I was doing this. 

So without DI, then we have these ugly things:
```
$entity_type_manager = \Drupal::entityTypeManager()->getStorage('field_config')
```
and 
```
$utils = \Drupal::service('islandora.utils');
```
Is this what you were referring to with _"just making static functions in the property class as opposed to injecting"_? I'm fine with that.

"Implementing my own static factory method" sounds hard. :( 

"plugin itself to pass in the requested values directly as opposed to the property itself" I'm not sure I follow. Right now the plugin does
```
$properties[$fieldname] = new EntityReferenceWithUriProperty($definition, $this->entityTypeManager, $this->utils);
```
and the Property's `__construct()` function accepts that signature... OOOOHH you're saying we could maybe remove the Property's `__create()` function?

Reply:
> implementing [a] static factory method

Really isn't that difficult. As in, it's effectively the same as `ContainerFactoryPluginInterface`, just without being passed the container... so similarly have a static method like the `::create()` it described, and the use the `\Drupal` global class dealio to access things from the container. So, something like:

```php
class EntityReferenceWithUriProperty {
  [...]
  public static function createInstance(array $definition) {
    return new static(
      $definition,
      // \Drupal::entityTypeManager() could probably equally work, anyway.
      \Drupal::service('entity_type.manager'),
      \Drupal::service('islandora.utils')
    );
  }
}
```

Sure, it might does still do a bit of binding with Drupal proper; however, the bigger issue is allowing the the services to be passed in the constructor that this would address (facilitating testability, should it be pursued), but doing so without requiring invoking classes to know what needs to be passed to the constructor other than what they _need_ to know (like the `$definition`, in this case).

Then, instead of the `new` to create, you can just hit this new static factory method, with something like: 

```php
$properties[$fieldname] = EntityReferenceWithUriProperty::createInstance($definition);
```

Reply:
On another note, looks like there should be schema definitions for any properties...

An example of those in `search_api` proper: https://git.drupalcode.org/project/search_api/-/blob/8.x-1.x/config/schema/search_api.processor.schema.yml#L220-271

Where they seem to be targeted, from the index config entity: https://git.drupalcode.org/project/search_api/-/blob/8.x-1.x/config/schema/search_api.index.schema.yml#L40-41

Might get a little messy given what I think would correspond to the key there, dynamic property names as per `$fieldname` over at https://github.com/Islandora/islandora/pull/913/files#diff-a6e10f5851d0df552f0c1523868a930cc3aac263d2c54582694ff5f979977bccR123-R124 ?

Reply:
Another potential alternative to deal with DI things could be to push off the building out of the lists of things another level, to define `FormElement`s that are aware of what they can display (under the assumption that `FormElement` plugins can work with the `ContainerFactoryPluginInterface` themselves), then the "form building" in the config form shouldn't need to have things injected at all?